### PR TITLE
`TypedStreamWriter` that writes to buffer

### LINKED ISSUE #9780: Implement `TypedStreamWriter` that writes to a `ByteBuffer`
- [ ] Copy the content from TypedStreamWriterImpl to a new class
- [ ] Remove the LogStreamBatch usage from that class, write directly into a pre-claimed buffer

### PR DESCRIPTION
**Pull request by pihme**
## Description

- Implements `TypedStreamWriter` that writes to buffer
- Implements processing result builder to use the buffered stream writer
- does not add tests. I started it, but it would take too much time. Hope I can revisit it after my holidays. For now there is just a follow up issue: https://github.com/camunda/zeebe/issues/9838

## Related issues

closes #9780 

## Definition of Done

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedProcessingResultBuilder.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  @Override
+  public ProcessingResult build() {
+    throw new RuntimeException("Not yet implemented");
+  }
+
+  private ValueType initValueType(final RecordValue value) {
+    final ValueType valueType = typeRegistry.get(value.getClass());
+    if (valueType == null) {
+      // usually happens when the record is not registered at the TypedStreamEnvironment
+      throw new RuntimeException("Missing value type mapping for record: " + value.getClass());
```
Comment (npepinpe) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedProcessingResultBuilder.java:92]:
```suggestion
      throw new IllegalArgumentException("Missing value type mapping for record: " + value.getClass());
```

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1: 

Looks good, I like the linear structure much more than the builder approach (since we anyway require all fields to be filled out).

- :x: One concern I have is related to the `canWriteAdditionalEvent` method. The functionality here is that there are processors where we want to handle, during processing, if writing an additional event will cause the write to fail downstream (e.g. job batch activation). Before, we were (indirectly) querying the dispatcher directly, but now we're back to using the `maxFragmentSize`. `maxFragmentSize` is unfortunately not the actual limit, since the dispatcher will frame entries as we add them. So if you have one entry, the real max fragment size for this writer would be `maxFragmentSize - frameLength`. But if you have multiple entries, it's `maxFragmentSize - (entryCount * frameLength)`. Actually, it's a little less than that, as we then align the size of the batches as well, which may cause them to require more memory than expected. So the real `maxFragmentSize` is more of a function of the number of entries and the total length of the batch. Anyway, all this to say, I think this currently breaks pre-existing functionality. We're now reverting to the implementation that it was before we fixed #5525 (relevant PR is #8799). Possibly we would be re-opening #5525 here. One problem is we never managed to write a good integration/regression test for #5525, so I don't know if we'd catch it now.

**Comment by pihme** [comment]
> * Anyway, all this to say, I think this currently breaks pre-existing functionality

I fear the same. I don't really see how we can avoid this, tbh. Do you have any idea how to solve this problem? (e.g. secure upper bounds; more precise calculations; try to write it regardless of capacity, then handle the error if and when it arises)

Without an approach, I am afraid this is a blocker for this development strain for now.

**Comment by npepinpe** [comment]
Well, in a perfect world, we wouldn't have any limitation associated with the dispatcher/follow up records. That's a much bigger scope, however. Maybe if instead of passing a static length we passed a function to compute it? It's definitely hacky without some major refactoring, but the current solution is already somewhat hacky.

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedProcessingResultBuilder.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    return valueType;
+  }
+
+  private BufferWriter initValueWriter(final RecordValue value) {
+    // validation
+    if (!(value instanceof BufferWriter)) {
```
[the remark this answers is not in the document]
Comment (pihme) [reply engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedProcessingResultBuilder.java:99]: No idea. Kept it as it was. Will add a todo, though

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedStreamWriter.java [hunk 3]
```diff
@@ -82,27 +84,24 @@ void appendRecord(
...
-    eventLength += metadataLength + valueLength;
-    eventCount += 1;
+    recordCount += 1;
   }
 
   boolean canWriteAdditionalEvent(final int length) {
-    final var count = eventCount + 1;
-    final var batchLength = eventLength + length + (count * HEADER_BLOCK_LENGTH);
-    return batchLength < maxFragmentSize;
+    return length < capacityCalculator.apply(recordBufferOffset);
```
Comment (npepinpe) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedStreamWriter.java:91]: :x: I expect the current function to map down to `Dispatcher#canClaimFragmentBatch(int recordCount, int batchLength)` at the end, so it should be a `BiFunction<Integer, Integer, Integer>`, I think.

**Review by npepinpe — CHANGES_REQUESTED** [review]

File: engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedStreamWriter.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  private final MutableDirectBuffer recordBuffer =
+      new ExpandableDirectByteBuffer(INITIAL_BUFFER_CAPACITY);
+
+  private final RecordMetadata metadata = new RecordMetadata();
+
+  private int recordBufferOffset;
+  private int recordCount;
+  private final BinaryOperator<Integer> capacityCalculator;
+
+  BufferedStreamWriter(final BinaryOperator<Integer> capacityCalculator) {
```
Comment (npepinpe) [inline engine/src/main/java/io/camunda/zeebe/streamprocessor/BufferedStreamWriter.java:41]: :+1: TIL about the `BinaryOperator` (though I shouldn't be surprised after the `UnaryOperator` :smile:)

**Review by npepinpe — APPROVED** [review]

**Comment by pihme** [comment]
bors merge

**Comment by pihme** [comment]
bors merge

**Comment by oleschoenburg** [comment]
@pihme test failures are legitimate, check out https://github.com/camunda/zeebe/runs/7427420911?check_suite_focus=true

**Comment by pihme** [comment]
> @pihme test failures are legitimate, check out https://github.com/camunda/zeebe/runs/7427420911?check_suite_focus=true

Thanks. I feel the heat is getting me

**Comment by oleschoenburg** [comment]
e8fb77b7792d5fd96501df765440b60fce119f65 (latest commit on this branch here) builds for me but when I rebase this branch on main it fails:
```
io.camunda.zeebe.streamprocessor.BufferedProcessingResultBuilder is not abstract and does not override abstract method resetPostCommitTasks() in io.camunda.zeebe.engine.api.ProcessingResultBuilder
```

**Comment by pihme** [comment]
bors merge

**Pull request merged**